From a9b3e2ddcdca7881d0bc5fd1f2eea502396c4308 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Tue, 10 Dec 2024 11:13:32 +0100
Subject: [PATCH 82/86] winhttp: Do not overrun buffer.

=================================================================
==956==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7f63049403bc at pc 0x6ffff7e1cf4a bp 0x7fffff8ff7e0 sp 0x7fffff8ff828
READ of size 2 at 0x7f63049403bc thread T-1
03d0:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffff7e1cf49 in add_request_headers /home/bernhard/data/entwicklung/2024/wine\wine/dlls/winhttp/request.c:500:16
    #1 0x6ffff7e1d3e2 in WinHttpAddRequestHeaders /home/bernhard/data/entwicklung/2024/wine\wine/dlls/winhttp/request.c:541:11
    #2 0x6ffff7ff55d5 in insert_http_header /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/msg.c:1912:9
    #3 0x6ffff7ff55d5 in message_insert_http_headers /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/msg.c:1985:15
    #4 0x6ffff7fd6825 in send_message_bytes /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/channel.c:1560:19
    #5 0x6ffff7fe2103 in send_message /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/channel.c:1731:10
    #6 0x6ffff7fe1e45 in send_message_proc /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/channel.c:1755:10
    #7 0x6ffff7fde5ae in queue_runner /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/channel.c:133:17
    #8 0x6ffffa903ad9 in tp_object_execute /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll\threadpool.c:2233:13
    #9 0x6ffffa90293f in threadpool_worker_proc /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll\threadpool.c:2364:13
    #10 0x6ffffbfd4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #11 0x6ffffa8afa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

0x7f63049403bc is located 0 bytes after 44-byte region [0x7f6304940390,0x7f63049403bc)
allocated by thread T0 here:
    #0 0x6ffff84ba2c1 in malloc /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:80:3
    #1 0x6ffff7ff5545 in build_http_header /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/msg.c:1899:18
    #2 0x6ffff7ff5545 in message_insert_http_headers /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/msg.c:1972:18
    #3 0x6ffff7fd6825 in send_message_bytes /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/channel.c:1560:19
    #4 0x6ffff7fe2103 in send_message /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/channel.c:1731:10
    #5 0x6ffff7fe1e45 in send_message_proc /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/channel.c:1755:10
    #6 0x6ffff7fde5ae in queue_runner /home/bernhard/data/entwicklung/2024/wine\wine/dlls/webservices/channel.c:133:17
    #7 0x6ffffa903ad9 in tp_object_execute /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll\threadpool.c:2233:13
    #8 0x6ffffa90293f in threadpool_worker_proc /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll\threadpool.c:2364:13
    #9 0x6ffffbfd4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #10 0x6ffffa8afa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/bernhard/data/entwicklung/2024/wine\wine/dlls/winhttp/request.c:500:16 in add_request_headers
Shadow bytes around the buggy address:
  0x7f6304940100: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
  0x7f6304940180: fa fa 00 00 00 00 00 02 fa fa fd fd fd fd fd fd
  0x7f6304940200: fa fa 00 00 00 00 06 fa fa fa 00 00 00 00 00 02
  0x7f6304940280: fa fa 00 00 00 00 00 fa fa fa 00 00 00 00 00 fa
  0x7f6304940300: fa fa 00 00 00 00 00 fa fa fa 00 00 00 00 00 02
=>0x7f6304940380: fa fa 00 00 00 00 00[04]fa fa fa fa fa fa fa fa
  0x7f6304940400: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7f6304940480: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7f6304940500: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7f6304940580: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7f6304940600: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==956==ABORTING
03d0:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
make: *** [Makefile:398531: dlls/webservices/tests/x86_64-windows/proxy.ok] Fehler 1
---
 dlls/winhttp/request.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winhttp/request.c b/dlls/winhttp/request.c
index cd3e4200e73..0a54e64d093 100644
--- a/dlls/winhttp/request.c
+++ b/dlls/winhttp/request.c
@@ -497,7 +497,7 @@ DWORD add_request_headers( struct request *request, const WCHAR *headers, DWORD
         for (q = p; q < headers + len && *q != '\r' && *q != '\n'; ++q)
             ;
         end = q;
-        while (*q == '\r' || *q == '\n')
+        while ((q < headers + len) && (*q == '\r' || *q == '\n'))
             ++q;
 
         if ((header = parse_header( p, end - p, FALSE )))
-- 
2.47.1

